core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -292,56 +286,44 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new LinkedList[T]()
+    private[this] var operations = Vector[T]()
 
-    def watched: Int = operations synchronized operations.size
+    def watched: Int = synchronized(operations.size)
 
     // add the element to watch
     def watch(t: T) {
-      operations synchronized operations.add(t)
+      synchronized {
+        operations = operations :+ t
+      }
     }
 
     // traverse the list and try to complete some watched elements
     def tryCompleteWatched(): Int = {
-
-      var completed = 0
-      operations synchronized {
-        val iter = operations.iterator()
-        while (iter.hasNext) {
-          val curr = iter.next()
-          if (curr.isCompleted) {
-            // another thread has completed this operation, just remove it
-            iter.remove()
-          } else if (curr synchronized curr.tryComplete()) {
-            completed += 1
-            iter.remove()
-          }
-        }
+      // call tryComplete without holding the lock to avoid potential deadlocks
+      var completedAlready = 0
+      var completedNow = 0
+      for (op <- synchronized(operations)) {
```

Comment:
Here's a sample of some runs with `TestPurgatoryPerformance` with varying message sizes. I held the following settings fixed: `--key-space-size 1000 --num 100000 --rate 1000 --pct50 20 --pct75 50 --timeout 100`.

First off of trunk:

| Request Size | Elapsed Time | Target Rate | Actual Rate | Process CPU Time | MarkSweep Count | MarkSweep Time | Scavenge Count | Scavenge Time |
| ------------- | -------------- | ----------- | ----------- | ------------------- | ------------------ | ----------------- | ------------------------ | --------------- |
| 10 | 99876 | 1002.234984 | 1002.224939 | 9707 | 0 | 4 | 0 | 34 |
| 100 | 99984 | 1001.191418 | 1001.191418 | 9669 | 0 | 4 | 0  | 31 |
| 1000 | 100192 | 999.020959 | 999.020959 | 8541 | 0 | 5 | 0 | 45 |
| 10000 | 100323 | 997.745096 | 997.735141 | 9891 | 0 | 11 | 0 | 77 |

Then with this patch:

| Request Size | Elapsed Time | Target Rate | Actual Rate | Process CPU Time | MarkSweep Count | MarkSweep Time | Scavenge Count | Scavenge Time |
| ------------- | -------------- | ----------- | ----------- | ------------------- | ------------------ | ----------------- | ------------------------ | --------------- |
| 10 | 100389 | 997.098444 | 997.088502 | 8841 | 0 | 4 | 0 | 34 |
| 100 | 100407 | 996.939396 | 996.939396 | 9416 | 0 | 3 | 0 | 21 |
| 1000 | 100490 | 996.105229 | 996.105229 | 9879 | 0 | 6 | 0 | 55 |
| 10000 | 100456 |  996.234235 | 996.234235 | 10138 | 0 | 11 | 0 | 75 |

The results do not show much difference. If anyone has recommendations for a better configuration to test against, please let me know and I will rerun.